File: v2/src/renderer/src/utils/ejson.ts
/**
 * Lightweight EJSON helpers that work on the POJO form returned by
 * JSON.parse() of a canonical EJSON string. The main process serializes
 * documents with `EJSON.stringify(doc, { relaxed: false })`, so e.g.
 * `ObjectId` becomes `{ "$oid": "..." }` and `Date` becomes
 * `{ "$date": "..." }`.
 */

export type EjsonValue =
  | string
  | number
  | boolean
  | null
  | EjsonValue[]
  | { [k: string]: EjsonValue };

export interface FormattedValue {
  /** Display label, e.g. `ObjectId("...")`. */
  label: string;
  /** Type tag for styling, e.g. `objectid`, `date`, `number`. */
  type:
    | 'string'
    | 'number'
    | 'boolean'
    | 'null'
    | 'objectid'
    | 'date'
    | 'long'
    | 'decimal'
    | 'binary'
    | 'regex'
    | 'minkey'
    | 'maxkey'
    | 'array'
    | 'object'
    | 'unknown';
}

export function parseRow(s: string): Record<string, EjsonValue> {
  return JSON.parse(s) as Record<string, EjsonValue>;
}

/**
 * Returns a primitive-style label + type for a leaf value. Object/array
 * containers return `array` / `object` so the caller can decide whether to
 * render a summary or recurse.
 */
export function formatValue(v: EjsonValue): FormattedValue {
  if (v === null) return { label: 'null', type: 'null' };
  if (typeof v === 'string') return { label: v, type: 'string' };
  if (typeof v === 'number') return { label: String(v), type: 'number' };
  if (typeof v === 'boolean') return { label: String(v), type: 'boolean' };

  if (Array.isArray(v)) {
    return { label: `[ ${v.length} ]`, type: 'array' };
  }

  // Object: check known EJSON wrappers
  if ('$oid' in v && typeof v.$oid === 'string') {
    return { label: `ObjectId("${v.$oid}")`, type: 'objectid' };
  }
  if ('$date' in v) {
    const raw = v.$date as EjsonValue;
    const value =
      typeof raw === 'string'
        ? raw
        : raw && typeof raw === 'object' && '$numberLong' in raw
          ? new Date(Number((raw as { $numberLong: string }).$numberLong)).toISOString()
          : '?';
    return { label: `ISODate("${value}")`, type: 'date' };
  }
  if ('$numberLong' in v) {
    return { label: `NumberLong(${(v as { $numberLong: string }).$numberLong})`, type: 'long' };
  }
  if ('$numberDecimal' in v) {
    return {
      label: `NumberDecimal("${(v as { $numberDecimal: string }).$numberDecimal}")`,
      type: 'decimal',
    };
  }
  if ('$binary' in v) {
    const binary = v.$binary as { base64?: string; subType?: string };
    return {
      label: `BinData(${binary.subType ?? '0'}, ${(binary.base64 ?? '').slice(0, 12)}…)`,
      type: 'binary',
    };
  }
  if ('$regularExpression' in v) {
    const r = v.$regularExpression as { pattern: string; options: string };
    return { label: `/${r.pattern}/${r.options ?? ''}`, type: 'regex' };
  }
  if ('$minKey' in v) return { label: 'MinKey()', type: 'minkey' };
  if ('$maxKey' in v) return { label: 'MaxKey()', type: 'maxkey' };

  const keys = Object.keys(v);
  return { label: `{ ${keys.length} field${keys.length === 1 ? '' : 's'} }`, type: 'object' };
}

export function isContainer(v: EjsonValue): v is EjsonValue[] | { [k: string]: EjsonValue } {
  return Array.isArray(v) || (typeof v === 'object' && v !== null);
}

export function prettyPrint(s: string): string {
  try {
    return JSON.stringify(JSON.parse(s), null, 2);
  } catch {
    return s;
  }
}

/**
 * Collects the union of top-level field names across a sample of rows.
 * Preserves first-seen order for stability.
 */
export function collectColumns(rows: string[]): string[] {
  const seen = new Set<string>();
  for (const row of rows) {
    try {
      const obj = JSON.parse(row) as Record<string, unknown>;
      for (const k of Object.keys(obj)) {
        if (!seen.has(k)) seen.add(k);
      }
    } catch {
      /* skip */
    }
  }
  return [...seen];
}

File: v3/src/main/kotlin/io/mex/util/Ejson.kt
package io.mex.util

import kotlinx.serialization.json.*

enum class EjsonType { String, Number, Boolean, Null, ObjectId, Date, Long, Decimal, Binary, Regex, MinKey, MaxKey, ArrayT, ObjectT, Unknown }

data class FormattedValue(val label: String, val type: EjsonType)

fun parseRow(s: String): JsonObject? = runCatching {
    Json.parseToJsonElement(s).jsonObject
}.getOrNull()

fun formatValue(v: JsonElement): FormattedValue = when (v) {
    is JsonNull -> FormattedValue("null", EjsonType.Null)
    is JsonPrimitive -> when {
        v.isString -> FormattedValue(v.content, EjsonType.String)
        v.booleanOrNull != null -> FormattedValue(v.content, EjsonType.Boolean)
        else -> FormattedValue(v.content, EjsonType.Number)
    }
    is JsonArray -> FormattedValue("[ ${v.size} ]", EjsonType.ArrayT)
    is JsonObject -> formatObject(v)
}

private fun formatObject(o: JsonObject): FormattedValue {
    if ("\$oid" in o) return FormattedValue("ObjectId(\"${(o["\$oid"] as JsonPrimitive).content}\")", EjsonType.ObjectId)
    if ("\$date" in o) {
        val raw = o["\$date"]!!
        val text = when (raw) {
            is JsonPrimitive -> raw.content
            is JsonObject -> (raw["\$numberLong"] as? JsonPrimitive)?.content?.let {
                java.time.Instant.ofEpochMilli(it.toLong()).toString()
            } ?: "?"
            else -> "?"
        }
        return FormattedValue("ISODate(\"$text\")", EjsonType.Date)
    }
    if ("\$numberLong" in o) return FormattedValue("NumberLong(${(o["\$numberLong"] as JsonPrimitive).content})", EjsonType.Long)
    if ("\$numberDecimal" in o) return FormattedValue("NumberDecimal(\"${(o["\$numberDecimal"] as JsonPrimitive).content}\")", EjsonType.Decimal)
    if ("\$binary" in o) {
        val b = o["\$binary"] as? JsonObject ?: return FormattedValue("BinData(…)", EjsonType.Binary)
        val sub = (b["subType"] as? JsonPrimitive)?.content ?: "0"
        val b64 = (b["base64"] as? JsonPrimitive)?.content ?: ""
        return FormattedValue("BinData($sub, ${b64.take(12)}…)", EjsonType.Binary)
    }
    if ("\$regularExpression" in o) {
        val r = o["\$regularExpression"] as? JsonObject ?: return FormattedValue("Regex(…)", EjsonType.Regex)
        val p = (r["pattern"] as? JsonPrimitive)?.content ?: ""
        val opt = (r["options"] as? JsonPrimitive)?.content ?: ""
        return FormattedValue("/$p/$opt", EjsonType.Regex)
    }
    if ("\$minKey" in o) return FormattedValue("MinKey()", EjsonType.MinKey)
    if ("\$maxKey" in o) return FormattedValue("MaxKey()", EjsonType.MaxKey)
    return FormattedValue("{ ${o.size} field${if (o.size == 1) "" else "s"} }", EjsonType.ObjectT)
}

fun isWrapper(o: JsonObject): Boolean =
    "\$oid" in o || "\$date" in o || "\$numberLong" in o || "\$numberDecimal" in o ||
        "\$binary" in o || "\$regularExpression" in o || "\$minKey" in o || "\$maxKey" in o

fun collectColumns(rows: List<String>): List<String> {
    val seen = LinkedHashSet<String>()
    for (row in rows) {
        parseRow(row)?.keys?.forEach { seen.add(it) }
    }
    return seen.toList()
}

fun prettyPrint(s: String): String = runCatching {
    val element = Json.parseToJsonElement(s)
    Json { prettyPrint = true; prettyPrintIndent = "  " }.encodeToString(JsonElement.serializer(), element)
}.getOrDefault(s)
